pkg/trait/environment.go

```diff
@@ -99,12 +102,85 @@ func (t *environmentTrait) Apply(e *Environment) error {
 		}
 	}
 
-	if t.Vars != nil {
-		for _, env := range t.Vars {
-			k, v := property.SplitPropertyFileEntry(env)
+	for _, env := range t.Vars {
+		k, v := property.SplitPropertyFileEntry(env)
+		switch {
+		case strings.HasPrefix(v, "configmap:"):
+
+			err := setValFromConfigMapKeySelector(&e.EnvVars, k, v)
+			if err != nil {
+				return err
+			}
+		case strings.HasPrefix(v, "secret:"):
+			err := setValFromSecretKeySelector(&e.EnvVars, k, v)
+			if err != nil {
+				return err
+			}
+		default:
 			envvar.SetVal(&e.EnvVars, k, v)
 		}
 	}
+	return nil
+}
 
+func setValFromConfigMapKeySelector(vars *[]corev1.EnvVar, envName string, path string) error {
+	vs, err := v1.DecodeValueSource(path, "", "invalid configmap reference: "+path)
```

Comment:
Sorry, I only gave you the link to the struct without explaining how I was thinking to use it. If you see, the object is already in charge to decode a string of type `configmap|secret:my-val`. The idea is to leverage such an abstraction by passing the user value without worrying about the type. Then, we can iterate over such objects and again, just setting `corev1.ConfigMapKeySelector` if such value is not nil, or `corev1.SecretMapKeySelector` accordingly. This would reduce the need to have these `setValFromXYZKeySelector` funcs

Reply:
Can this really be optimised without loosing specific error messages and the override of potentially existing EnvVarSource for a given key? I don't really understand why the `if envVar := envvar.Get(*vars, envName); envVar != nil` check is necessary. Should we get rid of that?

These setValFromXYZKeySelector funcs are mainly there for better readability. I could put the code in the above switch but there isn't any redundant duplication here afaict.

Reply:
You don't need to provide specialized error messages. This can be much more compacted (hence more maintainable and less error prone) by:
1. If not configmap/secret, just do the normal logic
2. if configmap/secret, decode the parameter (note that it returns an error we can bubble up, the operator would know what to do with it)
3. if the decoded value is a configmap (check its decoded struct is not nil), add it as a configmap, otherwise add as a secret
I don't see the need to verify if the env var already exists either. If it exists by any chance, we are just setting it again.